src/libraries/System.Diagnostics.DiagnosticSource/src/System/Diagnostics/DiagnosticSourceEventSource/DiagnosticSourceEventSourceActivitySourceListener.cs

```diff
@@ -0,0 +1,268 @@
+            Debug.Assert(currentActivitySourceSpec.SampleFunc != null);
+
+            DiagnosticSourceEventSourceFilterAndTransform? nextRaw = currentActivitySourceSpec.Next;
+
+            if (currentActivitySourceSpec.SourceName == "*")
+            {
+                AddCurrentToSpecs(currentActivitySourceSpec, _wildcardActivitySourceSpecs ??= new());
+            }
+            else
+            {
+                var sources = _activitySourceSpecsBySourceName ??= new(StringComparer.OrdinalIgnoreCase);
+
+                if (!sources.TryGetValue(currentActivitySourceSpec.SourceName, out ActivitySourceFilterAndTransformSpecs? specs))
+                {
+                    specs = new();
+                    sources[currentActivitySourceSpec.SourceName] = specs;
+                }
+
+                AddCurrentToSpecs(currentActivitySourceSpec, specs);
+            }
+
+            currentActivitySourceSpec = nextRaw;
+        }
+
+        Debug.Assert(_wildcardActivitySourceSpecs != null || _activitySourceSpecsBySourceName != null);
+
+        static void AddCurrentToSpecs(DiagnosticSourceEventSourceFilterAndTransform currentRaw, ActivitySourceFilterAndTransformSpecs specs)
+        {
+            if (currentRaw.ActivityName != null)
+            {
+                Dictionary<string, DiagnosticSourceEventSourceFilterAndTransform> specsByActivityName = specs.SpecsByActivityName ??= new(StringComparer.OrdinalIgnoreCase);
+
+                currentRaw.Next = !specsByActivityName.TryGetValue(currentRaw.ActivityName, out DiagnosticSourceEventSourceFilterAndTransform? head)
+                    ? null
+                    : head;
+
+                specsByActivityName[currentRaw.ActivityName] = currentRaw;
+            }
+            else
+            {
+                currentRaw.Next = specs.WildcardSpecs;
+                specs.WildcardSpecs = currentRaw;
+            }
+        }
+    }
+
+    private void CreateActivityListener()
+    {
+        Debug.Assert(_activityListener == null);
+        Debug.Assert(_wildcardActivitySourceSpecs != null
+            || _activitySourceSpecsBySourceName != null);
+
+        _activityListener = new ActivityListener();
+
+        _activityListener.SampleUsingParentId = OnSampleUsingParentId;
+        _activityListener.Sample = OnSample;
+
+        _activityListener.ShouldListenTo = (activitySource) =>
+        {
+            return _wildcardActivitySourceSpecs != null
+                || (_activitySourceSpecsBySourceName != null
+                && _activitySourceSpecsBySourceName.ContainsKey(activitySource.Name));
+        };
+
+        _activityListener.ActivityStarted = OnActivityStarted;
+
+        _activityListener.ActivityStopped = OnActivityStopped;
+
+        ActivitySource.AddActivityListener(_activityListener);
+    }
+
+    private void OnActivityStarted(Activity activity)
+    {
+        if (_wildcardActivitySourceSpecs?.OnActivityStarted(activity) == true)
```

Comment:
@noahfalk

I've pushed changes which I think/hope resolve everything 🤣

* Removed the ability to set a wildcard activity source + specific activity name rule (`[AS]*+MyActivityName`). This helped simplify things.

* Removed the ability to set multiple rules at the same level. Previously you could do things like `[AS]MySource/Start-Propagate\r\n [AS]MySource/Stop-Record`. Now only the first rule will win. This also helped simplify things.

* I implemented something like "TryGetBestSpec". However I didn't implement the caching of the result for a given source name + activity name. I'm a bit hesitant to do that because I'm worried activity name could be high cardinality. Some user _could_ do something silly like use a GUID for the activity name. Having unconstrained cache worried me. I do a bit of organization up front so the lookup is reasonably fast without cache.

* Reduced the abstractions. Now there is only...

  DiagnosticSourceEventSource
  DsesActivitySourceListener
  DsesFilterAndTransform
  DsesSamplerBuilder

  These nested helper types become top-level things:
  DsesActivityEvents
  DsesSampleActivityFunc